File: 프로그래머스/2/92342. 양궁대회/양궁대회.js
function solution(n, info) {
    var answer = [-1];
    
    let max = 0;
    let ryan = new Array(info.length).fill(0)
    
    function dfs(level, arrow){
        // 기저 조건
        if(level===11 || arrow ===0){
            if (arrow > 0) ryan[10] += arrow;
            
            let sum=0;
            for(let i=0; i<info.length;i++){
                if(ryan[i] > info[i]){          // ryan[i] <= info[i] 이건 둘 다 0점일 경우도 포함되기 때문에 안됌
                    sum+=10-i;
                }
                else if (info[i]){
                    sum-=10-i;
                }
            }
            
            if(sum>0 && sum>max){
                max = sum;
                answer = [...ryan]
            }
            else if(sum>0 && sum===max){
                for(let i=10; i>=0; i--){
                   if(ryan[i]!==answer[i]){
                       if(ryan[i]>answer[i]){
                           max=sum;
                           answer = [...ryan]
                       }
                       break;
                   }
                    
                }
            }
            
            if (arrow > 0) ryan[10] -= arrow; 
            return;
        }
        
        
        if(arrow>info[level]){
            ryan[level]=info[level]+1;
            dfs(level+1, arrow-ryan[level])
            ryan[level] = 0;
        }
        
        dfs(level+1, arrow)
    }
    
    dfs(0, n)
    
    return answer;
}

File: 프로그래머스/2/92342. 양궁대회/양궁대회.py
def solution(n, info):
    answer = []
    max_score = -1
    
    def dfs(cur, arrow, idx):
        nonlocal answer, max_score
        
        if idx == 11 or arrow == 0:
            temp = cur[:]
            if arrow > 0:
                temp[10] += arrow
            
            result = cal_score(info, temp)
            if result > 0 and max_score < result:
                max_score = result
                answer = temp[:]
            elif result > 0 and max_score == result:
                if temp[::-1] > answer[::-1]:
                    answer = temp[:]
            return
                
        need = info[idx] + 1
        
        if arrow >= need:
            cur[idx] = need
            dfs(cur, arrow - need, idx + 1)
            cur[idx] = 0
        
        dfs(cur, arrow, idx + 1)
        
    dfs([0] * 11, n, 0) 
    
    return answer if max_score > -1 else [-1]

def cal_score(apeach, ryan):
    a = 0
    r = 0
    
    for i in range(len(apeach)):
        if apeach[i] == 0 and ryan[i] == 0:
            continue
            
        if apeach[i] >= ryan[i]:
            a += 10 - i
        else:
            r += 10 - i
    
    return r - a